### ISSUE TITLE
error: TS2339 [ERROR]: Property 'gpu' does not exist on type 'Navigator'.

### ISSUE THREAD

**Issue by goddtriffin** [denoland/deno#22742]
Version: Deno 1.41.1

---

In Deno 1.8, [unstable/experimental support for WebGPU was added](https://deno.com/blog/v1.8#experimental-support-for-the-webgpu-api).
In Deno 1.39, [unstable/experimental support for WebGPU was re-added](https://deno.com/blog/v1.39#webgpu-is-back).

With WebGPU landing in Chrome Stable 113, I decided it was about time I start my foray into graphics programming. I do understand that [no other browser](https://caniuse.com/webgpu) has this feature enabled by default (yet).

Here is a basic example I tried getting up and running (stolen from the 1.39 re-announcement):

```typescript
// Try to get an adapter from the user agent.
const adapter = await navigator.gpu.requestAdapter();
if (adapter) {
  // Print out some basic details about the adapter.
  const adapterInfo = await adapter.requestAdapterInfo();

  // On some systems this will be blank...
  console.log(`Found adapter: ${adapterInfo.device}`);

  // Print GPU feature list
  const features = [...adapter.features.values()];
  console.log(`Supported features: ${features.join(", ")}`);
} else {
  console.error("No adapter found");
}
```

Here is the command I'm using to bundle my code: `deno bundle ui/static/script/run.ts bin/static/script/run.js`

And here is the associated `deno.jsonc` (make note - I'm targeting the browser):
```jsonc
{
    "compilerOptions": {
        "lib": [
            "dom",
            "dom.iterable",
            "dom.asynciterable",
            "deno.ns"
        ]
    },
    "unstable": [
        "webgpu"
    ],
    "lint": {
        "include": [
            "ui/static/script/"
        ]
    },
    "fmt": {
        "include": [
            "ui/static/script/"
        ],
        "lineWidth": 120
    },
}
```

When I try bundling this basic example, I get this error:
```
⚠️ Warning: `deno bundle` is deprecated and will be removed in Deno 2.0.
Use an alternative bundler like "deno_emit", "esbuild" or "rollup" instead.
Check file:///<repo>/ui/static/script/run.ts
error: TS2339 [ERROR]: Property 'gpu' does not exist on type 'Navigator'.
const adapter = await navigator.gpu.requestAdapter();
                                ~~~
    at file:///<repo>/ui/static/script/run.ts:2:33
```

I've tried a couple things:
1. Adding `--unstable` after `deno bundle`
   - (Yes - I'm aware `deno bundle` may be deprecated soon. The Deno community did not like this decision, and it seems like it is [still up for debate](https://github.com/denoland/deno/issues/11073#issuecomment-1605918145).)
2. Adding `"unstable": ["webgpu"]` to `deno.jsonc`.

Both of which seem to have zero affect. 

How do I get access to WebGPU when developing Typescript with Deno?

---

StackOverflow: https://stackoverflow.com/questions/76561043/deno-bundle-property-gpu-does-not-exist-on-type-navigator-ts2339

**Comment by goddtriffin** [comment]
Looks like I may have to import the types manually? `import * as _webgpu from "https://esm.sh/@webgpu/types";`

Why would this be required?

**Comment by 0f-0b** [comment]
TypeScript's `dom` library only includes features [supported by at least two browsers](https://github.com/microsoft/TypeScript-DOM-lib-generator#why-is-my-fancy-api-still-not-available-here). It is possible though to use Deno's WebGPU typings by adding `deno.webgpu` to the `lib` array in `deno.jsonc` and inserting the following declaration to `ui/static/script/run.ts`.

```ts
declare global {
  interface Navigator {
    readonly gpu: GPU;
  }
}
```

**Comment by goddtriffin** [comment]
When I (1) remove `import * as _webgpu from "https://esm.sh/@webgpu/types";` and (2) add `deno.webgpu` to `deno.jsonc`'s `lib` array, I get these errors:
```
error: TS2339 [ERROR]: Property 'gpu' does not exist on type 'Navigator'.
    if (!navigator.gpu) {
                   ~~~
    at file:///<repo>/ui/static/script/run.ts:33:20

TS2339 [ERROR]: Property 'gpu' does not exist on type 'Navigator'.
    const adapter: GPUAdapter | null = await navigator.gpu.requestAdapter();
                                                       ~~~
    at file:///<repo>/ui/static/script/run.ts:37:56

TS2322 [ERROR]: Type 'RenderingContext | null' is not assignable to type 'GPUCanvasContext | null'.
  Type 'CanvasRenderingContext2D' is missing the following properties from type 'GPUCanvasContext': configure, unconfigure, getCurrentTexture
    this.context = canvas.getContext("webgpu");
    ~~~~~~~~~~~~
    at file:///<repo>/ui/static/script/run.ts:43:5

Found 3 errors.
```

Then (3) when I add this:
```typescript
declare global {
  interface Navigator {
    readonly gpu: GPU;
  }
}
```
I get this error:
```
error: TS2322 [ERROR]: Type 'RenderingContext | null' is not assignable to type 'GPUCanvasContext | null'.
  Type 'CanvasRenderingContext2D' is missing the following properties from type 'GPUCanvasContext': configure, unconfigure, getCurrentTexture
    this.context = canvas.getContext("webgpu");
    ~~~~~~~~~~~~
    at file:///<repo>/ui/static/script/run.ts:49:5
```

I've modified my script (again, meant for the browser) to use some more WebGPU features.

I'm worried that I'm going to have to basically re-`declare` the entire WebGPU spec if I keep going down this path, which is what I assume the `import * as _webgpu from "https://esm.sh/@webgpu/types";` import is for...

**Comment by goddtriffin** [comment]
When I attempt running `deno run --check ui/static/script/run.ts` (with the same 3 changes as described in the previous comment), I get the same error:
```
error: TS2322 [ERROR]: Type 'RenderingContext | null' is not assignable to type 'GPUCanvasContext | null'.
  Type 'CanvasRenderingContext2D' is missing the following properties from type 'GPUCanvasContext': configure, unconfigure, getCurrentTexture
    this.context = canvas.getContext("webgpu");
    ~~~~~~~~~~~~
    at file:///<repo>/ui/static/script/run.ts:49:5
```

**Comment by goddtriffin** [comment]
When I remove the `declare global` and add back in the `@webgpu/types` import, everything compiles, bundles, and executes in the browser just fine.

I'm hoping there's a way to avoid manually declaring all these types and/or importing them. If the only way is to wait for Firefox/Safari to land WebGPU support and thus enable Typescript to land the types in the mainline branch, I understand. I'm technically unblocked by importing the types for now.

**Comment by Hajime-san** [comment]
It seems that the `Navigator` API belongs `deno.window`.

https://github.com/denoland/deno/blob/e1fb174f86adce421ee6bbce70e5dc1558c10868/cli/tsc/dts/lib.deno.window.d.ts#L108-L109

So, I think that a current workaround is like below.

```jsonc
{
    "compilerOptions": {
        "lib": [
            "dom",
            ...
            "deno.webgpu",
            // override previous types
            "deno.window"
        ]
    },
    ...
}
```

related issue: https://github.com/denoland/deno/issues/21981

**Comment by goddtriffin** [comment]
I tried removing the `@webgpu/types` import and adding `"deno.window"` to the bottom of my `deno.jsonc`'s `#/compilerOptions/lib` list and got this error:
```
error: TS2322 [ERROR]: Type 'RenderingContext | null' is not assignable to type 'GPUCanvasContext | null'.
  Type 'CanvasRenderingContext2D' is missing the following properties from type 'GPUCanvasContext': configure, unconfigure, getCurrentTexture
    const gpuCanvasContext: GPUCanvasContext | null = canvas.getContext("webgpu");
          ~~~~~~~~~~~~~~~~
    at file:///<repo>/ui/static/script/run.ts:55:11
```

**Comment by Hajime-san** [comment]
Hm, I can't understand your whole situation, but it's wonder that using `canvas`.
CanvasElement API is designed for browser, and Deno runtime don't have it.
I think that WebGPU on Deno don't need `canvas` like below example.

https://github.com/denoland/webgpu-examples/blob/fafa4a7a45fe8caf7c990c2fe5ea610c28b53004/hello-triangle/mod.ts

**Comment by goddtriffin** [comment]
@littledivy Your comment led me down a rabbit hole that helped me resolve this!

> This seems to be deno bundle specific

I've been meaning to migrate off of the soon-to-be-deprecated `deno bundle` for a long time now and this was the final push I needed. After admittedly too long, I finally figured out how to replace the bundling logic with `esbuild`. I wrote about my solution here: https://www.toddgriffin.me/blog/how-to-bundle-deno-typescript-for-the-browser

Anyways, after migrating to `esbuild` I removed the `@webgpu/types` import and I got no errors!

Here is my final config:
```jsonc
{
    "compilerOptions": {
        "lib": [
            "dom",
            "dom.iterable",
            "dom.asynciterable",
            "deno.ns",
            "deno.webgpu"
        ]
    },
    "unstable": [
        "webgpu"
    ]
}
```

@Hajime-san Luckily I never had to append `"deno.window"` to the compiler options!
